> ### **Assignment 2 - Numpy Array Operations** 
>
> This assignment is part of the course ["Data Analysis with Python: Zero to Pandas"](http://zerotopandas.com). The objective of this assignment is to develop a solid understanding of Numpy array operations. In this assignment you will:
> 
> 1. Pick 5 interesting Numpy array functions by going through the documentation: https://numpy.org/doc/stable/reference/routines.html 
> 2. Run and modify this Jupyter notebook to illustrate their usage (some explanation and 3 examples for each function). Use your imagination to come up with interesting and unique examples.
> 3. Upload this notebook to your Jovian profile using `jovian.commit` and make a submission here: https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas/assignment/assignment-2-numpy-array-operations
> 4. (Optional) Share your notebook online (on Twitter, LinkedIn, Facebook) and on the community forum thread: https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575 . 
> 5. (Optional) Check out the notebooks [shared by other participants](https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575) and give feedback & appreciation.
>
> The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.
>
> Try to give your notebook a catchy title & subtitle e.g. "All about Numpy array operations", "5 Numpy functions you didn't know you needed", "A beginner's guide to broadcasting in Numpy", "Interesting ways to create Numpy arrays", "Trigonometic functions in Numpy", "How to use Python for Linear Algebra" etc.
>
> **NOTE**: Remove this block of explanation text before submitting or sharing your notebook online - to make it more presentable.


# My 5 preferred Numpy Functions

NumPy library contgains various future which includes; a powerful N-dimensional array object, broadcasting function,and is useful in linear algebra, Fourier transform, and random numbers capabilities. 

- numpy.reshape ()
- numpy.split ()
- numpy.average ()
- numpy.linalg.det ()
- numpy.full ()

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [67]:
!pip install jovian --upgrade -q

In [68]:
import jovian

In [69]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Updating notebook "josephatkariuki24/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/josephatkariuki24/numpy-array-operations


'https://jovian.ai/josephatkariuki24/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [70]:
import numpy as np

In [71]:
# List of functions explained 
function1 = np.reshape
function2 = np.split
function3 = np.average
function4 = np.linalg.det
function5 = np.full

## Function 1 - np.reshape

The np.reshape () function shapes the array without changing the data of the array. This means to reshape a matrix, the new shape should be compatible with the original shape.

In [72]:
# Example 1 - shape array with two rows and five columns 
arr1 = np.arange(10).reshape(2,5)
print("\narray reshaped with 2 rows and 5 columns :\n" ,arr1)


array reshaped with 2 rows and 5 columns :
 [[0 1 2 3 4]
 [5 6 7 8 9]]


The array has been reshaped in form of 2 rows and 5 columns.

In [73]:
# Example 2 - constructing of a 3D array
arr2 = np.arange(27).reshape(3, 3, 3)
print("\nOriginal array reshaped to 3D : \n:", arr2)


Original array reshaped to 3D : 
: [[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]]


Reshaped array to a 3D array

In [74]:
# Example 3 
arr3 = np.arange(8).reshape(2,5)
print(arr3)

ValueError: cannot reshape array of size 8 into shape (2,5)

The above array broke because original array has 8 elements in total while the new array does not match with the old array length.To fixed it, the reshape array should be change in order to match the arr3 . The reshape should change from (2, 5) to (2, 4).

np.reshape can also be used to change from one dimension to another i.e change the number of elements in each dimension.

In [ ]:
jovian.commit()

## Function 2 - np.split

This function divides the array into subarrays along a specified axis

In [ ]:
#Example 1 - workings
arr1=np.arange(12)
print ("First array")
print (arr1)
print ('\n')

print("Split the array in 3 equal-sizes subarrays:")
arr2=np.split(arr1,3)
print (arr2)
print ('\n')




The above arrays has been divided into three equal part each array containing 4 elements.

In [ ]:
# Example 2 - split arra
arr2= np.split(arr1, 2, axis=0)

print(arr2)

The array above has been divided into two equal subarrays along the axis=0

In [ ]:
print(arr1)
arr2= np.split(arr1, (2,3), axis=1)

print(arr2)

In this case the tuple index is out of range ie it can not contain a number which is bigger than the maximum valued allowed by the data type. It can be fixed by adjusting the tuple to be within the range.

It is used to split an array function into more than one sub-arrays vertically(row-wise) and horizontally ie column wise

In [ ]:
jovian.commit()

## Function 3 - np.average
It is used to calculate the weighted average along the specified axis.

In [ ]:
# Example 1 
fibre = np.arange(1, 11)
fibre
np.average(fibre)

The answer is a result of summation of (1,2,3,4,5,6,7,8,9,10) then dividing it by 10 which the number of elements between range 1 to 11.

In [ ]:
# Example 2 
np.average(np.arange(2,7), weights=np.arange(6, 1, -1))

taking in considaration an array [2,3,4,5,6] and the corresponding weight [6,5,4,3,2] the weighted average is calculated by adding the product of corresponding elements and dividing the sum by the sum of weights.
weighted average = (2*6+3*5+4*4+5*3+6*2)/(6+5+4+3+2)

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
fibre = np.arange(6).reshape((3,2))
fibre

In [ ]:
np.average(fibre, weights=[1./5, 4./5])

the above function break because the shape of fibre and it's weights differ hence it can be fixed by specifying the axis eg axis=1

The np.average is used when the length of 1D weights is not the same as the shape of a long axis

In [75]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "josephatkariuki24/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/josephatkariuki24/numpy-array-operations


'https://jovian.ai/josephatkariuki24/numpy-array-operations'

## Function 4 - np.linalg.det

The np.linalg.det () is used to determine a square matrix's determinant. np.linalg.det () function takes an array as an arguement and return the determinat of a given array.

In [76]:
# Example 1 - working of linalg.det() in 2*2 matrix:
arr = np.arange(3, 7).reshape(2,2)
print("The array is :\n", arr)
print("The shape of the array is : ", np.shape(arr))

#we will use print determinant by using det() function
print("Determinant of the given array:", np.linalg.det(arr))


The array is :
 [[3 4]
 [5 6]]
The shape of the array is :  (2, 2)
Determinant of the given array: -1.9999999999999971


The determinant of the above array is -1.9999 which when you round of to the nearest whole number is 2. This can been calculated by; (3*6)-(4*5) which gives a determinant of -2

In [77]:
# Example 2 - working of linalg.det() in 3*3 matrix
arr2 = np.array([[7,3,1], [5, -3, 1], [2,6,9]]) 
print("Array is :\n", arr2)
print("The shape of the array is :", np.shape(arr2))

# Using determinant by using det() function
print("The determinant of the given array:",np.linalg.det(arr2))

Array is :
 [[ 7  3  1]
 [ 5 -3  1]
 [ 2  6  9]]
The shape of the array is : (3, 3)
The determinant of the given array: -323.9999999999999


The above answer -323.99 can be round of to whole number ie -324 which then can be calculated manually by carrying out the following operation; 7*(-3*9 -1*6) - 3*(5*9 - 1*2) + 1*(5*6 - -3*2) 

In [78]:
# Example 3 - breaking (to illustrate when it breaks)
arr3 = np.array([[3, 2, 1], [4, 5, 6]])
print("Array is :\n", arr3)
print("The shape of the array is :",np.shape(arr3))

#use np.linalg.det()
np.linalg.det(arr3)


Array is :
 [[3 2 1]
 [4 5 6]]
The shape of the array is : (2, 3)


LinAlgError: Last 2 dimensions of the array must be square

The above matrix function broke because the matrix has no equal number of raws and column ie it is not a square matrix. It can be fix by adding another raw so as to make it 3 by 3 matrix.

This function is used to calculate the determinant of the squared matrix.

In [ ]:
jovian.commit()

## Function 5 - np.full

full() function is used to return a new array of a given shape and data type filled with fill_value

In [ ]:
# Example 1 - creating array with two raws and two column

np.full([2,2],100)

In the above example I have created an array with two column and two raws.

In [79]:
# Example 2 - working
np.full((7, 3), (100, 10000, 1000000))

array([[    100,   10000, 1000000],
       [    100,   10000, 1000000],
       [    100,   10000, 1000000],
       [    100,   10000, 1000000],
       [    100,   10000, 1000000],
       [    100,   10000, 1000000],
       [    100,   10000, 1000000]])

In the above example , I have create an array with 7 rows and 3 columns.

In [80]:
# Example 3 - breaking (to illustrate when it breaks)
np.full((7,3),(100, 1000))

ValueError: could not broadcast input array from shape (2,) into shape (7,3)

In this example above we try to create an array with three columns and seven rows, but because we only set two numbers in each row it will not work. If we would like to fix this issue, we have to add one more number to the second parantheses.

It allow the creation of arrays which are not Numpy arrays

In [81]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "josephatkariuki24/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/josephatkariuki24/numpy-array-operations


'https://jovian.ai/josephatkariuki24/numpy-array-operations'

## Conclusion

NumPy array provides an array objects that is 50* faster than the traditional list. ndarry is the array object in NumPy which provides a lot of supporting function that makes working with ndarry very easy. Arrays are normally used in Data science where speed and resourses are very useful. 

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* ...

In [ ]:
jovian.commit()

<IPython.core.display.Javascript object>